dll/ntdll/nt_0600/ldr/ldrnotify.c

```diff
@@ -0,0 +1,151 @@
+
+static RTL_CRITICAL_SECTION_DEBUG LdrpDllNotificationLockDebug = {
+    .CriticalSection = &LdrpDllNotificationLock
+};
+
+static RTL_CRITICAL_SECTION LdrpDllNotificationLock = {
+    &LdrpDllNotificationLockDebug,
+    -1,
+    0,
+    0,
+    0,
+    0
+};
+
+NTSTATUS
+NTAPI
+LdrRegisterDllNotification(
+    _In_ ULONG Flags,
+    _In_ PLDR_DLL_NOTIFICATION_FUNCTION NotificationFunction,
+    _In_opt_ PVOID Context,
+    _Out_ PVOID *Cookie)
+{
+    PLDR_DLL_NOTIFICATION_ENTRY NewEntry;
+
+    /* Check input parameters */
+    if (Flags != 0 || NotificationFunction == NULL || Cookie == NULL)
+    {
+        return STATUS_INVALID_PARAMETER;
+    }
+
+    /* Allocate new entry and assign input values */
+    NewEntry = RtlAllocateHeap(LdrpHeap, 0, sizeof(*NewEntry));
+    if (NewEntry == NULL)
+    {
+        return STATUS_NO_MEMORY;
+    }
+    NewEntry->Callback = NotificationFunction;
+    NewEntry->Context = Context;
+
+    /* Add node to the end of global list */
+    RtlEnterCriticalSection(&LdrpDllNotificationLock);
+    if (LdrpDllNotificationList.Blink->Flink != &LdrpDllNotificationList)
+    {
+        FatalListEntryError(LdrpDllNotificationList.Blink->Blink,
+                            LdrpDllNotificationList.Blink,
+                            &LdrpDllNotificationList);
+    }
+    InsertTailList(&LdrpDllNotificationList, &NewEntry->List);
+    RtlLeaveCriticalSection(&LdrpDllNotificationLock);
+
+    /* Cookie is address of the new entry */
+    *Cookie = NewEntry;
+    return STATUS_SUCCESS;
+}
+
+NTSTATUS
+NTAPI
+LdrUnregisterDllNotification(
+    _In_ PVOID Cookie)
+{
+    NTSTATUS Status = STATUS_DLL_NOT_FOUND;
+    PLIST_ENTRY Entry;
+
+    /* Find entry to remove */
+    RtlEnterCriticalSection(&LdrpDllNotificationLock);
+    for (Entry = LdrpDllNotificationList.Flink;
+         Entry != &LdrpDllNotificationList;
+         Entry = Entry->Flink)
+    {
+        if (Entry == Cookie)
+        {
+            RtlpCheckListEntry(Entry);
+            RemoveEntryList(Entry);
+            Status = STATUS_SUCCESS;
+            break;
+        }
+    }
+    RtlLeaveCriticalSection(&LdrpDllNotificationLock);
+
+    if (NT_SUCCESS(Status))
+    {
+        RtlFreeHeap(LdrpHeap, 0, Entry);
+    }
+    return Status;
+}
+
+VOID NTAPI LdrpSendDllNotifications(
+    _In_ PLDR_DATA_TABLE_ENTRY DllEntry,
+    _In_ ULONG NotificationReason)
+{
+    PLIST_ENTRY Entry;
+    PLDR_DLL_NOTIFICATION_ENTRY NotificationEntry;
+    LDR_DLL_NOTIFICATION_DATA NotificationData;
+
+    /*
+     * LDR_DLL_LOADED_NOTIFICATION_DATA and LDR_DLL_UNLOADED_NOTIFICATION_DATA
+     * currently are the same, use _STATIC_ASSERT to ensure it then fill either of them.
+     */
+#define LdrpAssertDllNotificationDataMember(x)\
+    _STATIC_ASSERT(FIELD_OFFSET(LDR_DLL_NOTIFICATION_DATA, Loaded.x) ==\
+                   FIELD_OFFSET(LDR_DLL_NOTIFICATION_DATA, Unloaded.x))
+
+    _STATIC_ASSERT(sizeof(NotificationData.Loaded) == sizeof(NotificationData.Unloaded));
```

Comment:
`C_ASSERT` . Why not doing these elsewhere: outside of this `LdrpSendDllNotifications` function, maybe where the `LDR_DLL_NOTIFICATION_DATA` structure is defined?